field/subfields.go

```diff
@@ -0,0 +1,264 @@
+package field
+
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"sort"
+
+	"github.com/moov-io/iso8583/encoding"
+	"github.com/moov-io/iso8583/padding"
+)
+
+var _ Field = (*Subfields)(nil)
+
+// Subfields is a wrapper object designed to hold ISO8583
+// subfields. It packs and unpacks subfields ordered by ID.
+// The following are not supported by Subfields:
+//  - Fields that implement the field.Struct interface (see
+//    Subelements for such functionality).
+//  - Padding or encoding (responsibility for this is delegated
+//    recursively to the subfields themselves).
+type Subfields struct {
+	spec *Spec
+
+	orderedFieldIDs []int
+	idToFieldMap    map[int]Field
+
+	data interface{}
+}
+
+// NewSubfields creates a new instance of the *Subfield struct,
+// validates and sets its Spec before returning it.
+// Refer to SetSpec() for more information on Spec validation.
+func NewSubfields(spec *Spec) *Subfields {
+	f := &Subfields{}
+	f.SetSpec(spec)
+	return f
+}
+
+// Spec returns the receiver's spec.
+func (f *Subfields) Spec() *Spec {
+	return f.spec
+}
+
+// SetSpec validates the spec and creates new instances of Fields defined
+// in the specification.
+// NOTE: Subfields do not support encoding and padding. Therefore, users should
+// only pass None or nil values for these types. Passing any other value will
+// result in a panic. In addition, all fields defined within the *Spec must
+// not implement the field.Struct interface.
+func (f *Subfields) SetSpec(spec *Spec) {
+	if err := validateSubfieldsSpec(spec); err != nil {
+		panic(err)
+	}
+	f.spec = spec
+	f.idToFieldMap = spec.CreateMessageFields()
+	f.orderedFieldIDs = orderedKeys(f.idToFieldMap)
+}
+
+// SetData uses reflection to traverse through fields provided in the data
+// parameter and does the following:
```

Comment:
I don't think we need to say `uses reflection` as we don't need to expose implementation detail. These comments are super-helpful though